### PR TITLE
Elastic Beanstalk Web Server [predict labels & self maintenance]

### PR DESCRIPTION
**Pull request by CathyZhang0822**
Migrate PR 11935 here.

## Description ##
An elastic beanstalk web server which uses ML models to predict labels

@marcoabreu @larroy @Roshrini @lanking520 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] Code is well-documented: 
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by lanking520** [comment]
Hi Cathy please follow this guideline to remove all `DS_STORE` https://stackoverflow.com/questions/107701/how-can-i-remove-ds-store-files-from-a-git-repository

**Comment by lanking520** [comment]
Please also remove intelliJ configure file... `.idea`

File: mxnet-bot/LabelBotPredict/elastic_beanstalk_server/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+WORKDIR /src
+
+#install Python modules
+RUN pip install -r requirements.txt
+
+# Run it
+ENTRYPOINT ["python", "application.py"]
+
+# Environment Variables
+ENV github_user your_github_id
```
Comment (lanking520) [inline mxnet-bot/LabelBotPredict/elastic_beanstalk_server/Dockerfile]: Could you please set these before entry point?

File: mxnet-bot/LabelBotPredict/elastic_beanstalk_server/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+matplotlib==2.2.2
+MarkupSafe==1.0
+nltk==3.3
+numpy==1.14.5
+pandas==0.23.3
+pyparsing==2.2.0
+python-dateutil==2.7.3
+pytz==2018.5
+requests==2.19.1
+s3transfer==0.1.13
```
Comment (lanking520) [inline mxnet-bot/LabelBotPredict/elastic_beanstalk_server/requirements.txt]: Still, are we using S3 here?

File: mxnet-bot/LabelBotPredict/label_bot_send_predictions/LabelBot.py [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+import datetime
+import operator
+import os
+import boto3
+import re
+import logging
+logging.basicConfig(level=logging.INFO)
+
+
+class LabelBot:
```
Comment (lanking520) [inline mxnet-bot/LabelBotPredict/label_bot_send_predictions/LabelBot.py]: Email Bot Ver2? Not LabelBot, please change the name. Label Recommender?

File: mxnet-bot/LabelBotPredict/elastic_beanstalk_server/Dockerfile (hunk 1 above)
Comment (CathyZhang0822) [reply mxnet-bot/LabelBotPredict/elastic_beanstalk_server/Dockerfile]: Done

File: mxnet-bot/PredictLabels/application.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+import timeit
+import atexit
+import logging
+import os.path
+
+logging.basicConfig(level=logging.INFO)
+
+application = Flask(__name__)
+
+if not os.path.exists('/tmp/Classifier.p'):
```
Comment (larroy) [inline mxnet-bot/PredictLabels/application.py]: Can we do something better than these hardcoded files? A directory to save the model which is used as a parameter to those guys? Then you can use tempfile.TemporaryDirectory  in the script to select the one for this particular run. If it has to be shared, needs to be a runtime argument or env variable.

File: mxnet-bot/PredictLabels/README.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+requirements.txt stopwords.txt
+```
+It will zip all needed files into `eb.zip`
+* Manually create a new Elastic Beanstalk application.
+    1. Go to AWS Elastic Beanstalk console, click ***Create New Application***. Fill in *Application Name* and *Description*, click ***Create***.
+    2. Under ***Select environment tier***, select ***Web server environment***, click ***Select***.
+    3. Under **Base configuration**, select **Preconfigured platform**. In its dropdown, select **Docker**. Then select ***Upload your code***, upload `eb.zip`.
+    4. Click ***Configure more options***. Modify Intances, in the dropdown of Instance type, select t2.large. Click ***Create Environment*** (No need to select a security group, EB will create one.)
+    5. It will take about 2 minutes to setup the environment. 
+    6. Once the environment is setup, it will take 5-10 minutes to generate models.
```
Comment (larroy) [inline mxnet-bot/PredictLabels/README.md]: Are there logs collected somewhere? is the output redirected to CloudWatch?

File: mxnet-bot/PredictLabels/SentenceParser.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    ssl._create_default_https_context = _create_unverified_https_context
+import os.path
+import pandas as pd
+import re
+import sys
+import logging
+
+logging.basicConfig(level=logging.INFO)
+
+# English Stopwords
```
Comment (larroy) [inline mxnet-bot/PredictLabels/SentenceParser.py]: Can you move it to the class body (not init) if you want it to be a class variable?

File: mxnet-bot/PredictLabels/Trainer.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+# This script is served to train Machine Learning models
+from DataFetcher import DataFetcher
+from SentenceParser import SentenceParser
+from sklearn.feature_extraction.text import TfidfVectorizer
+from sklearn.svm import SVC
+from sklearn.naive_bayes import GaussianNB
+from sklearn.preprocessing import LabelEncoder
+import pickle
+import logging
+logging.basicConfig(level=logging.INFO)
```
Comment (larroy) [inline mxnet-bot/PredictLabels/Trainer.py]: this should be set only in the main file, not in every module.

File: mxnet-bot/PredictLabels/application.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+from apscheduler.triggers.interval import IntervalTrigger
+from Predictor import Predictor
+from Trainer import Trainer
+import plot_piechart
+import timeit
+import atexit
+import logging
+import os.path
+
+logging.basicConfig(level=logging.INFO)
```
Comment (larroy) [inline mxnet-bot/PredictLabels/application.py]: same as above

File: mxnet-bot/PredictLabels/application.py (hunk 4 above)
Comment (larroy) [inline mxnet-bot/PredictLabels/application.py]: same comment about hardcoded path

File: mxnet-bot/PredictLabels/test_trainer.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+							  					"number":11924,
+							 					"labels":[{'name':'Bug'}],
+							  					"state":"closed",
+							  					"title":"Issue in exporting gluon model",
+							  					"html_url":"https://github.com/apache/incubator-mxnet/issues/11924",
+							  				  }]
+			self.trainer.train()
+
+if __name__ == "__main__":
+	unittest.main()
```
Comment (larroy) [inline mxnet-bot/PredictLabels/test_trainer.py]: please add newline

File: mxnet-bot/PredictLabels/test_datafetcher.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+							  					"state":"closed",
+							  					"title":"issue's title",
+							  					"html_url":"https://github.com/apache/incubator-mxnet/issues/11925",
+							  				  }]
+			self.df.data2json('all', labels=["Feature"], other_labels=False)
+			expected_data = [{'id': 11925, 'title': "issue's title", 'body': "issue's body", 'labels': 'Feature'}] 						 
+			self.assertEqual(expected_data, self.df.json_data)
+
+if __name__ == "__main__":
+	unittest.main()
```
Comment (larroy) [inline mxnet-bot/PredictLabels/test_datafetcher.py]: newline missing

File: mxnet-bot/PredictLabels/README.md (hunk 5 above)
Comment (CathyZhang0822) [reply mxnet-bot/PredictLabels/README.md]: No logs are not in CloudWatch. In Elastic Beanstalk console, click "Logs". Then you can check it.

File: mxnet-bot/PredictLabels/SentenceParser.py (hunk 6 above)
Comment (CathyZhang0822) [reply mxnet-bot/PredictLabels/SentenceParser.py]: moved stopwords into class

File: mxnet-bot/PredictLabels/Trainer.py (hunk 7 above)
Comment (CathyZhang0822) [reply mxnet-bot/PredictLabels/Trainer.py]: Now I only set logging's level in application.py.

File: mxnet-bot/PredictLabels/test_trainer.py (hunk 9 above)
Comment (CathyZhang0822) [reply mxnet-bot/PredictLabels/test_trainer.py]: newline added

**Comment by CathyZhang0822** [comment]
I made a force push to squash my 2nd commit "add .gitignore". It seems that some conversations were erased. Here are Pedro's review:
1. In `application.py`, line 35 - 38.  Don't hard code `if not os.path.exists('/tmp/Classifier.p'):` here.
2. Only set logging's level in main file
3. In `SentenceParse.py`. put english stopwords in class.
4. Add new lines to `test_trainer.py`, `test_datafetcher.py`

I revised my code and those changes were merged in 1st commit

File: mxnet-bot/PredictLabels/DataFetcher.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+        """
+        This method is to count how many pages of issues/labels in total
+        state can be "open"/"closed"/"all"
+        """
+        url = 'https://api.github.com/repos/%s/issues' % self.repo
+        response = requests.get(url, {'state': state},
+                                auth=self.auth)
+        assert response.status_code == 200, "Authorization failed"
+        if "link" not in response.headers:
+            return 1
```
Comment (larroy) [inline mxnet-bot/PredictLabels/DataFetcher.py]: why do we return 1 here?  shouldn't we raise an exception instead?

File: mxnet-bot/PredictLabels/Trainer.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        Y = le.fit_transform(labels)
+        # Step4: Train Classifier
+        # SVC, kernel = 'rbf'
+        logging.info("Training Data..")
+        # clf = SVC(gamma=0.5, C=100, probability=True)
+        clf = self.clf
+        clf.fit(X, Y)
+        # Step5: save models
+        logging.info("Saving Models..")
+        pickle.dump(tv, open("/tmp/Vectorizer.p", "wb"))
```
Comment (larroy) [inline mxnet-bot/PredictLabels/Trainer.py]: Can we do something with these hardcoded paths? not sure if you missed my previous comments on tmpdir

File: mxnet-bot/PredictLabels/Predictor.py [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+        """
+        DF = DataFetcher()
+        df_test = DF.fetch_issues(issues)
+        rule_based_predictions = []
+        for i in range(len(issues)):
+            # extract every issue's title
+            row = df_test.loc[i, 'title']
+            # apply rule-based algorithms
+            single_issue_predictions = []
+            if "feature request" in row.lower(): single_issue_predictions.append("Feature")
```
Comment (larroy) [inline mxnet-bot/PredictLabels/Predictor.py]: break line please

File: mxnet-bot/PredictLabels/Predictor.py [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+            if "c++" in row.lower(): single_issue_predictions.append("C++")
```
Comment (larroy) [inline mxnet-bot/PredictLabels/Predictor.py]: break line

File: mxnet-bot/PredictLabels/Trainer.py (hunk 12 above)
Comment (CathyZhang0822) [reply mxnet-bot/PredictLabels/Trainer.py]: Didn't miss it. Planned to ask more details in today's office hour.

File: mxnet-bot/PredictLabels/Predictor.py [hunk 15]
```diff
@@ -41,15 +42,14 @@ def __init__(self):
...
 
-    def reload(self):
+    def reload(self, tv_file, clf_file, labels_file):
         """
         This method is to load models
         """
-        self.tv = pickle.load(open("/tmp/Vectorizer.p", "rb"))
-        self.labels = pickle.load(open("/tmp/Labels.p", "rb"))
-        self.clf = pickle.load(open("/tmp/Classifier.p", "rb"))
+        self.tv = pickle.load(open(tv_file, "rb"))
```
Comment (larroy) [inline mxnet-bot/PredictLabels/Predictor.py]: Use "with open" idiom please, this leaks fds.

File: mxnet-bot/PredictLabels/Trainer.py [hunk 16]
```diff
@@ -21,29 +21,48 @@
...
+    labels = ["Performance", "Test", "Question",
+               "Feature request", "Call for contribution",
+               "Feature", "Example", "Doc",
+               "Installation", "Build", "Bug"]
 
-    def __init__(self):
+    def __init__(self, 
+                 tv=TfidfVectorizer(min_df=0.00009, ngram_range=(1, 3), max_features=10000), 
+                 clf=SVC(gamma=0.5, C=100, probability=True),
+                 tmp_tv_file=tempfile.NamedTemporaryFile(),
```
Comment (larroy) [inline mxnet-bot/PredictLabels/Trainer.py]: Why don't we put all of them in a temporary directory? I think this would be cleaner. The file names can be hardcoded in this temporary directory.

Comment (CathyZhang0822) [reply mxnet-bot/PredictLabels/Trainer.py]: Done. Put all of them in a tempfile.TemporaryDirectory

File: mxnet-bot/PredictLabels/Predictor.py (hunk 15 above)
Comment (CathyZhang0822) [reply mxnet-bot/PredictLabels/Predictor.py]: Done. Replace all "open" with "with open",
